autogen/agentchat/conversable_agent.py

```diff
@@ -29,6 +29,145 @@ def colored(x, *args, **kwargs):
 F = TypeVar("F", bound=Callable[..., Any])
 
 
+class _Register:
+    """A class for decorators registering functions to be used by an agent.
+
+    The decorator factory is returned by the `for_execution` and `for_llm` methods.
+    """
+
+    def __init__(self, agent: "ConversableAgent"):
+        """Initialize the Register class.
+
+        Args:
+            agent: the agent to be registered.
+        """
+        self._agent = agent
+
+    def for_execution(
+        self,
+        *,
+        name: Optional[str] = None,
+    ) -> Callable[[F], F]:
+        """Decorator factory for registering a function to be executed by an agent.
+
+        It's return value is used to decorate a function to be registered to the agent.
+
+        Args:
+            name (optional(str)): name of the function. If None, the function name will be used (default: None).
+
+        Returns:
+            The decorator for registering a function to be used by an agent.
+
+        Examples:
+            ```
+            @user_proxy.register.for_execution()
+            @agent2.register_for_llm()
+            @agent1.register_for_llm(description="This is a very useful function")
+            def my_function(a: Annotated[str, "description of a parameter"] = "a", b: int, c=3.14):
+                 return a + str(b * c)
+            ```
+
+        """
+        return self._agent.register_for_execution(name=name)
+
+    def for_llm(
+        self,
+        *,
+        name: Optional[str] = None,
+        description: Optional[str] = None,
+    ) -> Callable[[F], F]:
+        """Decorator factory for registering a function to be used by an agent.
```

Comment:
To be used for LLM function or tool call geneation. Note this requires the LLM to support function or tool call following the OpenAI api specification.

Reply:
done 👍